### DISCUSSION TITLE
POCL with X86 host and RISCV target : compilation issue

### DISCUSSION THREAD

**Discussion by jeromefereyre** [pocl/pocl#1513] [Q&A] [answered]
Hi,

I'm currently working with a RISCV accelerator hosted in an FPGA (Xilinx connected through PCIe) and i'm trying to integrate it within POCL to offload kernels from the X86 host into this RISCV target.

I'm connecting pocl with an existing linux driver to copy argument and kernel from/to the accelerator.

I have some POCL flow that are already working yet, like copying kernel arguments into FPGA memory.

But i'm facing issues with the compilation of kernel.

It seems that the generation of the program.bc do not generate errors.
But when i call clang in my device specific pocl driver i encounter this kind of message

.../install/llvm/bin/clang -v  /tmp/pocl_xxx_program_nZ1XT4.bc   -o /tmp/pocl_xxx_program_vbJlMR.elf
...
error: Invalid record (Producer: 'LLVM17.0.1' Reader: 'LLVM 17.0.1')
1 error generated.

Does anybody has any recipe to build a working POCL environment able to run from X86 host and compile kernels for RISCV(64) target ?

Thanks for reading :) 
Jerome

**Comment by franz** [comment]
Hi,

Does `llvm-dis /tmp/pocl_xxx_program_nZ1XT4.bc` produce an error ? Also, it should be possible to validate a bitcode using `opt` with `verify` pass (i don't remember the exact cmdline). I've seen `error: Invalid record` before, but it was caused by version mismatch. In your case the versions are identical, so i suspect that invalid bitcode has been generated quietly somewhere (unfortunately this can happen..)

**Reply by jeromefereyre** [reply]
Hello,

Effectively, the lvm-dis command ends with the same error on the program.bc file.

llvm-dis /tmp/pocl_xxx_program_nZ1XT4.bc
../install/llvm/bin/llvm-dis: error: Invalid record (Producer: 'LLVM17.0.1' Reader: 'LLVM 17.0.1')

If i understand well, program.bc is some kind of aggregation beetween kernel-riscv64.bc and the kernel source file.
Just to be sure, the dis command returns no error on kernel-riscv64.bc.

So it seems that the problem is localized in program bitecode generation...
Is there a way to monitor this step ?

And the opt commands fails with the same error.

Thank you :)
Jé

**Reply by franz** [reply]
> If i understand well, program.bc is some kind of aggregation beetween kernel-riscv64.bc and the kernel source file.

Sort of. First OpenCL C source is compiled with `clang -emit-llvm` (though using the C++ API `clang::EmitLLVMOnlyAction`), and the bitcode is then (optionally) processed further.

The function `pocl_llvm_build_program` in `pocl_llvm_build.cc` takes care of calling Clang's emit-llvm action. If called with non-zero `linking_program` argument, it will additionally call `generateProgramBC` from the same file. That one currently does three things: 1) applies some treatment to printf calls 2) handles program scope variables 3) links the builtin library (`kernel-riscv64.bc` in your case). It also runs `llvm::verifyModule` if `POCL_LLVM_VERIFY` env variable is set.

You could try setting that env var to 1 to run the verification, and then move the call of `llvm::verifyModule` a few lines in `generateProgramBC` to see which step is causing the invalid bitcode.

**Reply by pjaaskel** [reply, marked as answer]
Just a hunch, but this could be linker.cpp calling some of the library functions using wrong arg types at the call site. These passed silently and manifested only as illegal bitcodes at some point when there was such an issue. The bitcode is still dumpable if you don't write it out so it can help to dump it in linker.cpp using the dump() method. Which OpenCL C builtins are you calling from your test kernel? 

By the way, are you using the AlmaIF driver for invoking your accelerator? It might be useful for your use case.

**Reply by jeromefereyre** [reply]
Thank you for your messages :)

I have moved step forward into the debugging of my issue by moving the verifyModule call.
And as you mentionned it was linked to a bad function call in the kernel that pass quite silently the link phase and ends with an invalid bytecode.

I found that by activating the DBPRINT macro in linker.cpp.

Concerning the accelerator invocation, i'm not using the Almaif driver.
In my configuration the bitstream is installed in FPGA through manual call to Xilinx tools (and do not change over time) and POCL driver interracts with a dedicated linux driver dealing with the accelerator.

May Almaif be relevant in this context ?

**Reply by pjaaskel** [reply]
Good that my hunch was right this time.

> May Almaif be relevant in this context ?

Yes, I think so. AlmaIF is documented primarily for invoking built-in-kernels, but it also has an interface for soft kernels. It seems the documentation is a bit lacking in this aspect, but @TopiLeppanen is the lead author and is a helpful guy.

**Reply by TopiLeppanen** [reply]
Yes, there is some potential to be useful.
I have been running custom ASIP processors on Xilinx SoC and PCIe FPGAs (and on Intel PCIe FPGAs, and our custom ISA simulator) via the AlmaIF-driver. The processor implements the memory-mapped interface as described in almaif.rst https://portablecl.org/docs/html/almaif.html.

I wrote some more about our compilation flow to the document in PR https://github.com/pocl/pocl/pull/1514
There's some info in that document on how we compile the kernels for OpenASIP, which could be relevant to your use case.
For even more info, there's the paper: https://doi.org/10.1016/j.micpro.2023.104772

The AlmaIF runtime driver uses the memory-mapped AlmaIF API to launch kernels and move data around. Internally the AlmaIF-driver then calls into e.g. Xilinx XRT driver.
